tez-common/src/main/java/org/apache/tez/util/LoggingUtils.java

```diff
@@ -0,0 +1,147 @@
+
+  @SuppressWarnings("unchecked")
+  public static void initLoggingContext(ThreadLocalMap threadLocalMap, Configuration conf,
+      String dagId, String taskAttemptId) {
+    Hashtable<String, String> data = (Hashtable<String, String>) threadLocalMap.get();
+    if (data == null) {
+      data = new NonClonableHashtable<String, String>();
+      threadLocalMap.set(data);
+    }
+    data.put("dagId", dagId == null ? "" : dagId);
+    data.put("taskAttemptId", taskAttemptId == null ? "" : taskAttemptId);
+
+    String[] mdcKeys = conf.getStrings(TezConfiguration.TEZ_MDC_CUSTOM_KEYS,
+        TezConfiguration.TEZ_MDC_CUSTOM_KEYS_DEFAULT);
+
+    if (mdcKeys.length == 0) {
+      return;
+    }
+
+    String[] mdcKeysValuesFrom = conf.getStrings(TezConfiguration.TEZ_MDC_CUSTOM_KEYS_CONF_PROPS,
+        TezConfiguration.TEZ_MDC_CUSTOM_KEYS_CONF_PROPS_DEFAULT);
+    LOG.info("MDC_LOGGING: setting up MDC keys: keys: {} / conf: {}", Arrays.asList(mdcKeys),
+        Arrays.asList(mdcKeysValuesFrom));
+
+    int i = 0;
+    for (String mdcKey : mdcKeys) {
+      // don't want to fail on incorrect mdc key settings, but warn in app logs
+      if (mdcKey.isEmpty() || mdcKeysValuesFrom.length < i + 1) {
+        LOG.warn("cannot set mdc key: {}", mdcKey);
+        break;
+      }
+
+      String mdcValue = mdcKeysValuesFrom[i] == null ? "" : conf.get(mdcKeysValuesFrom[i]);
+      // MDC is backed by a Hashtable, let's prevent NPE because of null values
+      if (mdcValue != null) {
+        data.put(mdcKey, mdcValue);
+      } else {
+        LOG.warn("MDC_LOGGING: mdc value is null for key: {}, config key: {}", mdcKey,
+            mdcKeysValuesFrom[i]);
+      }
+
+      i++;
+    }
+  }
+
+  public static String getPatternForAM(Configuration conf) {
+    return conf.get(TezConfiguration.TEZ_LOG_PATTERN_LAYOUT_AM, null);
+  }
+
+  public static String getPatternForTask(Configuration conf) {
+    return conf.get(TezConfiguration.TEZ_LOG_PATTERN_LAYOUT_TASK, null);
+  }
+
+  /**
+   * This method is for setting a NonClonableHashtable into log4j's mdc. Reflection hacks are
+   * needed, because MDC.mdc is well protected (final static MDC mdc = new MDC();). The logic below
+   * is supposed to be called once per JVM, so it's not a subject to performance bottlenecks. For
+   * further details of this solution, please check NonClonableHashtable class, which is set into
+   * the ThreadLocalMap. A wrong outcome of this method (any kind of runtime/reflection problems)
+   * should not affect the DAGAppMaster/TezChild. In case of an exception a ThreadLocalMap is
+   * returned, but it won't affect the content of the MDC.
+   */
+  @SuppressWarnings("unchecked")
+  public static ThreadLocalMap setupLog4j() {
+    ThreadLocalMap mdcContext = new ThreadLocalMap();
+    mdcContext.set(new NonClonableHashtable<String, String>());
+
+    try {
+      final Constructor<?>[] constructors = org.apache.log4j.MDC.class.getDeclaredConstructors();
```

Comment:
log4j or slf4j?

Reply:
log4j, intentionally, the whole hack implemented in this method is log4j specific